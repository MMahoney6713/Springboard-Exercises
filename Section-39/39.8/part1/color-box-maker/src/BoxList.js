import React from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { useState } from 'react';
import { v4 as uuid } from 'uuid';

const BoxList = () => {
    const [boxArray, setBoxArray] = useState([])

    const addBox = ({color, width, height}) => {
        setBoxArray(boxArray => ([
            ...boxArray,
            {
                id: uuid(),
                color: color,
                width: parseInt(width),
                height: parseInt(height)
            }
        ]))
    }

    const removeBox = (id) => {
        const newBoxArray = boxArray.filter(box => box.id !== id)
        setBoxArray(newBoxArray);
    }

    return (
        <div className='BoxList'>
            <NewBoxForm addBox={addBox}/>
            {boxArray.map(box => (
                <div key={box.id}>
                    <Box 
                        color={box.color} 
                        width={box.width} 
                        height={box.height} 
                    />
                    <button onClick={() => removeBox(box.id)}>X</button>
                </div>
            ))}
        </div>
    )
}

export default BoxList;