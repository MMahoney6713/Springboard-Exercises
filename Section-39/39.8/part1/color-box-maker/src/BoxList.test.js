import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from './BoxList';

it("renders without crashing", () => {
    render(<BoxList />)
})

it("matches snapshot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

it("adds a box when form is submitted", () => {
    const { getByLabelText, queryByText } = render(<BoxList />)

    expect(queryByText('X')).not.toBeInTheDocument();

    const colorInput = getByLabelText('Color:');
    const heightInput = getByLabelText('Height:');
    const widthInput = getByLabelText('Width:');
    const submitButton = queryByText('Add Box')
    
    fireEvent.change(colorInput, {target: {value: 'red'}});
    fireEvent.change(heightInput, {target: {value: '200'}});
    fireEvent.change(widthInput, {target: {value: '200'}});
    fireEvent.click(submitButton);

    expect(queryByText('X')).toBeInTheDocument();
})

it("removes box when remove box button is fired", () => {
    const { getByLabelText, queryByText } = render(<BoxList />)

    const colorInput = getByLabelText('Color:');
    const heightInput = getByLabelText('Height:');
    const widthInput = getByLabelText('Width:');
    const submitButton = queryByText('Add Box')
    
    fireEvent.change(colorInput, {target: {value: 'red'}});
    fireEvent.change(heightInput, {target: {value: '200'}});
    fireEvent.change(widthInput, {target: {value: '200'}});
    fireEvent.click(submitButton);

    expect(queryByText('X')).toBeInTheDocument();

    const removeBoxButton = queryByText('X');
    fireEvent.click(removeBoxButton);

    expect(queryByText('X')).not.toBeInTheDocument();
})