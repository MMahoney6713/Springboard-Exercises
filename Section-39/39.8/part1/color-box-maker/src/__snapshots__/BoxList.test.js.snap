// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="color"
      >
        Color: 
      </label>
      <input
        id="color"
        name="color"
        placeholder="color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        placeholder="width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        placeholder="height"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
